### ISSUE TITLE
Extension manager in 2.1 does not annotate Jupyter packages with Jupyter badges

### ISSUE THREAD

**Issue by jasongrout** [jupyterlab/jupyterlab#8161]
In the old extension manager, a whitelist of packages were annotated with a Jupyter logo and a note in the tooltip to indicate they were from Project Jupyter:

<img width="384" alt="Screen Shot 2020-04-02 at 11 10 41 PM" src="https://user-images.githubusercontent.com/192614/78329664-33be9300-7537-11ea-8974-5827019db205.png">

This seems to have disappeared with the new extension manager (note that the lab icon is just generically the npm icon, and has nothing to do with extra information to the user outside of the package contents).

<img width="455" alt="Screen Shot 2020-04-02 at 11 12 36 PM" src="https://user-images.githubusercontent.com/192614/78329814-77190180-7537-11ea-8a48-5d12ab8396c9.png">

Is this omission intentional?

**Comment by jasongrout** [comment]
CC @saulshanabrook @echarles

**Comment by echarles** [comment]
> Is this omission intentional?

Yes, this has been discussed in https://github.com/jupyterlab/jupyterlab/pull/8078#issuecomment-605214635 (read also and next comments).

Although the jlab icon is removed, the jlab org extensions are still sorted at the top of the list.

**Comment by jasongrout** [comment]
> Yes, this has been discussed in [#8078 (comment)](https://github.com/jupyterlab/jupyterlab/pull/8078#issuecomment-605214635) (read also and next comments).

Thanks. I didn't see in that discussion a reason why the extensions from Project Jupyter are not noted anymore (sorting isn't an explicit indicator, of course, unless there is at least some sort of division between the Project Jupyter extensions and other extensions).

@tgeorgeux, I gather from the discussion linked above that the mockup did not include an indication of which extensions are from Project Jupyter. Was there a reason to remove this indication?

**Comment by jasongrout** [comment]
On the other hand, reading more carefully, maybe @tgeorgeux didn't make a mockup...

@saulshanabrook, as near as I can tell, the reason was this comment from you:

> The "developed by project jupyter" icon is still there. Maybe we should remove this now that we have these other icons?

But IIRC, the icons in the new extension are from the project's own data, so it's not the same as the whitelist-type approach we had before, where we showed a Project Jupyter icon next to projects that *we* specifically are annotating.

**Comment by echarles** [comment]
The first mock on https://github.com/jupyterlab/jupyterlab/pull/8078#issue-393090186 still shows the icon, true.

I will open anyway a PR to address https://github.com/jupyterlab/jupyterlab/issues/8149

I can add in that PR the icon (waiting on confirmation for that).

BTW the previous method to show that icon was called isWithelisted` which I had renamed `isJupyterlabOrg` before removing it. This icon should not be confused with the new White listing feature. These are completely different features and there may be a risk to bring confusion.

**Comment by jasongrout** [comment]
First of all, I should have started with: *Thank you* for all the work you are doing on this extension manager. We *really* appreciate it!

I originally thought this indication was superseded by the whitelist feature. I think you are right that really this is a different issue, though.

**Comment by echarles** [comment]
Thx for your reviews and glad it is useful. There are still cycles before the release to rollback that. I am listening to the final decision.

**Comment by saulshanabrook** [comment]
> @saulshanabrook, as near as I can tell, the reason was this comment from you:

Sorry, I should have articulated what I was thinking.

It looked bad to me to have two icons, one small, one big, of the same thing. The little ones just looked off once we added the bigger ones. 

And for the user, the big ones are mostly a superset of the information expressed in the little icon. The only difference would be if an extension author changed their github photo to be the JupyterLab  photo, in which case a user might get confused.

But I think that's an edge case, intentionally malicious behavior, that we shouldn't be designing around. 🤷‍♂  If they are malicious extensions, that are trying to pretend like they are made by the Jupyter team, I think we could take other actions to remove them from that list...

**Comment by krassowski** [comment]
Cross-referencing my earlier comment on this https://github.com/jupyterlab/jupyterlab/issues/8043#issuecomment-599245780. I would prefer the icon indicator to stay, but I also admit that it does not seem visually appealing in the current format (having both a large badge AND an indicator, often both with the same logo). 

Would it make sense (not necessarily in this immediate release, but possibly later on) to use a different placement/kind of an indicator? A few quick ideas

- a ribbon in a corner (a dirty sketch below)
  ![idea](https://user-images.githubusercontent.com/5832902/78367054-614c2080-75b9-11ea-992b-26fa68fede93.png)
  <details>
  For a quick demo I used a mix of live editing & GIMP post-processing (to cut the corners), sharing relevant bits for others to expeiment with:

  ```html
   <div class="jp-ribbon">by Jupyter</div>
  ```

   ```css
  .jp-ribbon {
    background: #fc6f00;
    color: white;
    font-size: 10px;
    height: 15px;
    position: absolute;
    right: 0;
    transform: translate(20px,8px) rotate(45deg);
    padding: 1px 20px;
  }
  ```
  </details>
- an inverted logo in a corner, such as in [github-corners](https://github.com/tholman/github-corners) - would take less space
- just the logo in the corner
- just some text in the corner
  ![Screenshot from 2020-04-03 15-40-06](https://user-images.githubusercontent.com/5832902/78372849-b3913f80-75c1-11ea-9225-89ecbdb3a417.png)


A potentially wider question is whether users should have to opt-in (accept the warning) to install the extensions from JupyterLab org (`isJupyterlabOrg == true`).

**Comment by krassowski** [comment]
@saulshanabrook I agree with that the potentially malicious activity is an edge case (and it might be better to go with the 2.1 release as it), but as a user, I felt much reassured by this icon. Like if someone uploads `@jupylerab/toc` with the same icon and description...

**Comment by echarles** [comment]
> A potentially wider question is whether users should have to opt-in (accept the warning) to install the extensions from JupyterLab org (isJupyterlabOrg == true).

My feeling is that we should not overload user mind with conditions. 

To bring more confidence to the user, the real solution is the administrator to setup a whitelist...then talking about warning comes the question: should the user explicitly accept the warning if the admin has setup a whitelist, especially with the new text content proposed in https://github.com/jupyterlab/jupyterlab/pull/8162

**Comment by echarles** [comment]
> A potentially wider question is whether users should have to opt-in (accept the warning) to install the extensions from JupyterLab org (isJupyterlabOrg == true).

We should also define what are the extensions to receive that badge: Only under jupyterlab or all jupyter family, like e.g. `jupyter-voila` (https://github.com/voila-dashboards/voila/blob/master/packages/jupyterlab-voila/package.json#L2)

**Comment by jasongrout** [comment]
> To bring more confidence to the user, the real solution is the administrator to setup a whitelist...then talking about warning comes the question: should the user explicitly accept the warning if the admin has setup a whitelist, especially with the new text content proposed in #8162

Agreed - in fact, even moreso, I suppose some administrators would like to *not* give extra indication of Project Jupyter projects.

If we do still want to give indication, to resolve the double-icon issue, perhaps instead a (much clearer) new line of text under the description could be added: "Maintained by Project Jupyter"

**Comment by saulshanabrook** [comment]
> If we do still want to give indication, to resolve the double-icon issue, perhaps instead a new line of text under the description could be added: "Maintained by Project Jupyter". A new line of text would be much clearer anyway.

That sounds reasonable...

**Comment by jasongrout** [comment]
> And for the user, the big ones are mostly a superset of the information expressed in the little icon. The only difference would be if an extension author changed their github photo to be the JupyterLab photo, in which case a user might get confused.

Well, it's more than that.

1. Lab has a very non-standard Jupyter icon. If you didn't know the lab icon, you might think it was fake and made to look like Jupyter. It certainly doesn't follow the Jupyter branding guidelines, for example.
2. Things like voila, which are part of Project Jupyter, don't even have the Jupyter icon at all.

**Comment by jasongrout** [comment]
To be fair, though, we also provide all of *our* software with no warranty, liability, etc., and we should not imply that we do have any sort of extra assurance.

I think the thing other app stores have that we are missing is the concept of an organizational owner that extends beyond how the code is organized. I don't know how we would get that information reliably in general in our current npm-based approach.

**Comment by jasongrout** [comment]
In the interest of being *really* clear about where we are getting this extension information, how about having the extension name not linking to the package.json url entry, but instead to the npm page for the version displayed. From there the user can go to the url from the npm interface.

The reason for this is that any package can point to any url as their "url", so that is in a sense untrusted data. However, we can point directly to the actual npm page, which (a) isn't something a package can spoof, and (b) more clearly communicates that we are getting this extension data from npm, and (c) emphasizing that jlab extensions really are just special npm packages.

I'm not sure what that implies for using custom npm registry apis to get packages.

**Comment by jasongrout** [comment]
I suppose we have the same spoofing issues with how we automatically get the github org image from the package.json url entry.

**Comment by echarles** [comment]
I like the idea to link to npm and the first time I clicked on an extension, I have been surprised to fall on github instead of npm. If link to npm, I find weird to display the github account badge. Unfortunately, npm does not provide any badge (nor in the response, nor on the package page - https://www.npmjs.com/package/@jupyterlab/coreutils).

Regarding the spoofing risks, npm can also be spoofed btw.

**Comment by jasongrout** [comment]
Do you have a screenshot of what you are suggesting? 

On the other hand, I think we had just decided to forego the Jupyter badge annotation for now. Is that all right with you as well?

**Comment by jasongrout** [comment]
> If link to npm, I find weird to display the github account badge. Unfortunately, npm does not provide any badge (nor in the response, nor on the package page - https://www.npmjs.com/package/@jupyterlab/coreutils).

I agree it's a bit weird to pull our source info from npm, but then infer the icon to use from github. We could also roll back the icon for now until we have a wider discussion about the implications of it, and conservatively just list the npm package name and one-line description, as well as a link to the npm page for that specific version.

**Comment by jasongrout** [comment]
> Regarding the spoofing risks, npm can also be spoofed btw.

Can you elaborate?

**Comment by tgeorgeux** [comment]
>Do you have a screenshot of what you are suggesting?

The img above that Krassowski posted: 
![image](https://user-images.githubusercontent.com/22109376/78400150-03b0d780-75ab-11ea-9c0c-7cdf873a9565.png)
>On the other hand, I think we had just decided to forego the Jupyter badge annotation for now. Is that all right with you as well?

That is fine with me. I read through the comments and thought we were trying to get it in, I am ok with not having the badge for now or indefinitely. It just depends on what people find more helpful.

**Comment by echarles** [comment]
> > Regarding the spoofing risks, npm can also be spoofed btw.
> Can you elaborate?

If we want to question the risks that what we see is not what it is supposed, we can imagine a DNS attack that would redirect all our request to npm to a hacker.

**Comment by jasongrout** [comment]
> If we want to question the risks that what we see is not what it is supposed, we can imagine a DNS attack that would redirect all our request to npm to a hacker.

Ah. That is quite a bit more sophisticated than setting my url to some Jupyter repo.

**Comment by jasongrout** [comment]
> Showing github icons and linking to github sounds fine to me.

Another issue I see with linking to github is that you click on an extension and you see the code, which seems more intimidating than seeing the npmjs page.

**Comment by krassowski** [comment]
I would prefer users to go to the GitHub page as this is where they can report issues (and even be tempted to open a PR). I think that this works for GitLab and bitbucket as well (and that in an ideal world, GitLab pictures should be used as badges as well).

In general, I would like to have control over this by the `hompage` attribute in `package.json`, and I would be happy to accept a whitelist of sites that will be made into clickable links inside of JupyterLab extension manager. But if I understood correctly, package.json is not readily available before install... or is it? I remember some code to check version compatibility...

<details>

Also, it is not always super easy to maintain a nice NPM website (https://www.npmjs.com/package/@krassowski/jupyterlab-lsp), but I always find it easy to have a nice GitHub README. For example, the links on the NPM may not work as intended and it is even worse in a monorepo where having the readme in the first place requires some tricks (or maybe I am the issue here, IDK).

</details>

**Comment by jasongrout** [comment]
Can we have both? Can we have a link to the npmjs page, and then a separate link to the website it claims? Like:

[@jupyterlab/toc-extension](https://www.npmjs.com/package/@jupyterlab/toc) ([website](https://github.com/jupyterlab/jupyterlab-toc))

**Comment by jasongrout** [comment]
Also, is this content escaped when it is put on the page? Since it is coming from untrusted sources, we need to make sure it is being escaped.

**Comment by echarles** [comment]
@jasongrout @tgeorgeux @saulshanabrook @krassowski  I have opened https://github.com/jupyterlab/jupyterlab/pull/8168 to propose a change that hopefully takes into account discussions we all had here and will create a consensus.

**Comment by saulshanabrook** [comment]
Fixed in #8168
